### PR TITLE
[stats] Enable percentile-based capping (winsorization)

### PR DESCRIPTION
**Pull request by lukesonnet**
### Features and Changes

* Changes to `MetricInterface` to have two new fields, and commensurate changes to all places that touch capping (forms, integrations, APIs) and a JIT migration to handle the new `LegacyMetricInterface` which converts old capping into new, "absolute" capping
  * `capping` - either `null` for no capping, or "absolute" for the current absolute capping behavior, and "percentile" for the percentile based capping
  * `capValue` - the value for absolute capping or the percentile (0 to 1) to use
* Makes appropriate FE changes for metricsform
* Adds percentile computation to compute the percentile of user-level aggregates in the same experiment metric query to produce the least overhead in terms of managing a service that measures the percentiles separately.
* Returns `main_cap_value` and `denominator_cap_value` in the `View Queries` modal so people can at least glimpse what the capping value being used is.
* Works withe POST and PUT endpoints, where a user can turn off capping by setting it to `null`.

- Closes GB-270, GB-271

TODO:

- [x] consider returning cap value for transparency
- [x] additional analysis of computing the cap value post-experiment (this approach introduces bias, but requires no new data and we are already "biasing" data by capping)
- [x] validation that percentile capping value is (0, 1)

### Dependencies

n/a

### Testing

- [x] All engines approach matches (athena and databricks untested; note MS SQL uses approximation so it is different)
- [x] Test metrics API endpoints
- [x] Add unit test for migration
- [x] QA in app (see screenshots below)

Tested that create metric API endpoint works with just `cap` (backwards compatibility); with `capping` and `capValue` (as expected); and validates capping and capvalue both must exist (good validation, error messaging).

<details>
<summary>API endpoints work as expected (payloads, responses)</summary>

```
## backwards compatibility
# payload
{
    "datasourceId": "ds_2s2j1dlqlbr44wom",
    "type": "count",
    "name": "Orders metric in the API",
    "sql": {
      "identifierTypes": ["user_id"],
      "conversionSQL": "SELECT userid as user_id, timestamp as timestamp, 1 as value FROM sample.orders"
    },
    "behavior": {
      "cap": 10
    }
}

# response
{
  "metric": {
    "id": "met_2s2jxfiljg0bjl8",
    "name": "Orders metric in the API",
    "description": "",
    "dateCreated": "2023-06-28T17:45:07.820Z",
    "dateUpdated": "2023-06-28T17:45:07.820Z",
    "archived": false,
    "datasourceId": "ds_2s2j1dlqlbr44wom",
    "owner": "",
    "projects": [],
    "tags": [],
    "type": "count",
    "behavior": {
      "goal": "increase",
      "capping": "absolute",
      "capValue": 10,
      "minPercentChange": 0.005,
      "maxPercentChange": 50,
      "minSampleSize": 150,
      "riskThresholdDanger": 0.0125,
      "riskThresholdSuccess": 0.0025,
      "conversionWindowStart": 0,
      "conversionWindowEnd": 72
    },
    "sql": {
      "identifierTypes": [
        "user_id"
      ],
      "conversionSQL": "SELECT userid as user_id, timestamp as timestamp, 1 as value FROM sample.orders",
      "userAggregationSQL": "SUM(value)",
      "denominatorMetricId": ""
    }
  }
}


## as expected
# payload
{
    "datasourceId": "ds_2s2j1dlqlbr44wom",
    "type": "count",
    "name": "Orders metric in the API",
    "sql": {
      "identifierTypes": ["user_id"],
      "conversionSQL": "SELECT userid as user_id, timestamp as timestamp, 1 as value FROM sample.orders"
    },
    "behavior": {
      "capping": "percentile",
      "capValue": 0.99
    }
}

# response
{
  "metric": {
    "id": "met_2s2jxfiljg0bduk",
    "name": "Orders metric in the API",
    "description": "",
    "dateCreated": "2023-06-28T17:45:00.380Z",
    "dateUpdated": "2023-06-28T17:45:00.380Z",
    "archived": false,
    "datasourceId": "ds_2s2j1dlqlbr44wom",
    "owner": "",
    "projects": [],
    "tags": [],
    "type": "count",
    "behavior": {
      "goal": "increase",
      "capping": "percentile",
      "capValue": 0.99,
      "minPercentChange": 0.005,
      "maxPercentChange": 50,
      "minSampleSize": 150,
      "riskThresholdDanger": 0.0125,
      "riskThresholdSuccess": 0.0025,
      "conversionWindowStart": 0,
      "conversionWindowEnd": 72
    },
    "sql": {
      "identifierTypes": [
        "user_id"
      ],
      "conversionSQL": "SELECT userid as user_id, timestamp as timestamp, 1 as value FROM sample.orders",
      "userAggregationSQL": "SUM(value)",
      "denominatorMetricId": ""
    }
  }
}


## error (have to specify both)
# payload
{
    "datasourceId": "ds_2s2j1dlqlbr44wom",
    "type": "count",
    "name": "Orders metric in the API",
    "sql": {
      "identifierTypes": ["user_id"],
      "conversionSQL": "SELECT userid as user_id, timestamp as timestamp, 1 as value FROM sample.orders"
    },
    "behavior": {
      "capValue": 0.99
    }
}

# response
{
  "message": "Must specify both `behavior.capping` and `behavior.capValue or neither."
}


## error (have to specify percentile b/w 0-1
# payload
{
    "datasourceId": "ds_2s2j1dlqlbr44wom",
    "type": "count",
    "name": "Orders metric in the API",
    "sql": {
      "identifierTypes": ["user_id"],
      "conversionSQL": "SELECT userid as user_id, timestamp as timestamp, 1 as value FROM sample.orders"
    },
    "behavior": {
      "capping": "percentile",
      "capValue": 99
    }
}

# response
{
  "message": "When using percentile capping, `behavior.capValue` must be between 0 and 1."
}
```

</details>

### Screenshots

<details>
<summary>cap_value is returned properly in query view</summary>

Screenshot 2023-06-28 at 10 43 32 AM

</details>

<details>
<summary>Metrics form behaves as expected</summary>

https://www.loom.com/share/a8f23ac5a9314459a12ae49adf6d5946

</details>
<details>
<summary>no capping, pctile, absolute, and ratio capping all work side by side</summary>
Screenshot 2023-06-16 at 11 06 28 AM
</details>
<details>
<summary>deprecated config.yml import displays right error message</summary>
Screenshot 2023-06-16 at 11 06 28 AM
</details>
<details>
<summary>config.yml import/export works correctly</summary>

```
...
  met_2s2j1yf7livegbir:
    type: count
    name: Number of Purchases (BQ) (capped)
    datasource: ds_2s2j19hglfbpwq2r
    capping: percentile
    capValue: 0.75
...
```

</details>

### CODE REVIEW

File: packages/back-end/src/api/openapi/payload-schemas/PostMetricPayload.yaml [hunk 1]
```diff
@@ -47,7 +47,17 @@ properties:
           - decrease
       cap:
         type: number
-        description: This should be non-negative
+        description: (deprecated, use capping and capValue fields instead) This should be non-negative
```
Comment (tinahollygb) [inline packages/back-end/src/api/openapi/payload-schemas/PostMetricPayload.yaml:50]: Looks like according to the [OpenAPI docs](https://swagger.io/docs/specification/describing-parameters/) you can add `deprecated: true` and it should show up with proper styling for deprecated fields.

File: packages/back-end/src/services/experiments.ts [hunk 2]
```diff
@@ -1031,6 +1031,28 @@ export function postMetricApiPayloadIsValid(
...
+    const capValueExists = typeof capValue !== "undefined";
+    if (cappingExists !== capValueExists) {
+      return {
+        valid: false,
+        error:
+          "Must specify both `behavior.capping` and `behavior.capValue or neither.",
+      };
+    }
+    if (capping && capValue && capping === "percentile") {
+      if (capValue > 1) {
```
Comment (bryce-fitzsimons) [inline packages/back-end/src/services/experiments.ts:1048]: these 2 lines can be:
`if (capping === "percentile" && capValue > 1)`
or if capValue could be undefined, then:
`if (capping === "percentile" && (capValue || 0) > 1`

**Review by bryce-fitzsimons — APPROVED** [review]
I did not test this e2e, but I don't see any real risks in launching this. Everything looks good.

File: packages/front-end/components/Metrics/MetricForm/index.tsx [hunk 3]
```diff
@@ -948,16 +963,41 @@ const MetricForm: FC<MetricFormProps> = ({
...
+                  className="px-3 py-2 pb-0 mb-4 border rounded"
+                >
+                  <label>Capped Value</label>
+                  <input
+                    type="number"
+                    step="any"
+                    min="0"
+                    max={form.watch("capping") === "percentile" ? "1" : ""}
+                    className="form-control"
+                    {...form.register("capValue")}
```
[the remark this answers is not in the document]
Comment (lukesonnet) [reply packages/front-end/components/Metrics/MetricForm/index.tsx:992]: I use `min` and `max` directly on the `input` element, whereas elsewhere (e.g. for regressionAdjustmentDays) we use a `validate` function in the `register` call. Which is preferred for us?

[the remark this answers is not in the document]
Comment (jdorn) [reply packages/front-end/components/Metrics/MetricForm/index.tsx:992]: How you have it is correct

**Pull request merged**